unity-environment/Assets/ML-Agents/Examples/GridWorld/Scripts/GridAgent.cs

```diff
@@ -16,7 +16,26 @@ public override void InitializeAgent()
 
     public override void CollectObservations()
     {
-
+        // Prevents the agent from picking an action that would make it collide with a wall
+        var positionX = (int) transform.position.x;
+        var positionZ = (int) transform.position.z;
+        var maxPosition = academy.gridSize - 1;
+        if (positionX == 0)
```

Comment:
the constants 0 -> 3 in the conditionals (if position...) and in SetActionMask should be class static variables.